asf_core_data/pipeline/mcs/process/mcs_epc_joining.py

```diff
@@ -0,0 +1,392 @@
+        drop_epc_address (bool, optional): Whether or not to drop the address
+        from the EPC records used for matching. Defaults to True.
+        verbose (bool, optional): Whether or not to print diagnostic information.
+        Defaults to True.
+
+    Returns:
+        Dataframe: Matched MCS-EPC records.
+    """
+    if hps is None:
+        print("Getting HP data...")
+        hps = get_processed_installations_data()
+
+    if epcs is None:
+        epc_version = "preprocessed" if all_records else "preprocessed_dedupl"
+        print("Getting EPC data...")
+        # TODO: replace with line that loads most recent full EPC data
+        epcs = load_s3_data(
+            bucket_name,
+            "outputs/EPC/EPC_GB_preprocessed_and_deduplicated_sample_prox.csv",
+            usecols=None,
+        )
+
+    prepared_hps = prepare_hps(hps)
+    prepared_epcs = prepare_epcs(epcs)
+
+    joined = join_prepared_mcs_epc_data(
+        prepared_hps,
+        prepared_epcs,
+        all_records=all_records,
+        drop_epc_address=drop_epc_address,
+        verbose=verbose,
+    )
+
+    return joined
+
+
+def select_most_relevant_epc(joined_df):
+    """From a "fully joined" MCS-EPC dataset, chooses the "best"
+    EPC for each HP installation record (i.e. the one that is assumed
+    to best reflect the status of the property at the time of HP installation).
+    The EPC chosen is the latest one before the installation if it exists;
+    otherwise it is the earliest one after the installation.
+
+    Args:
+        joined_df (Dataframe): Joined MCS-EPC data. Assumed to
+        contain INSPECTION_DATE, commission_date and original_mcs_index columns.
+
+    Returns:
+        Dataframe: Most relevant MCS-EPC records.
+    """
+
+    # Sort joined_df by INSPECTION_DATE
+    joined_df = joined_df.sort_values("INSPECTION_DATE").reset_index(drop=True)
+
+    # Identify rows where the EPC data is before the MCS one
+    joined_df["epc_before_mcs"] = (
+        joined_df["INSPECTION_DATE"] <= joined_df["commission_date"]
+    )
+
+    # Identify indices of last EPC record before MCS
+    last_epc_before_mcs_indices = (
+        joined_df.reset_index()
+        .loc[joined_df["epc_before_mcs"]]
+        .groupby("original_mcs_index")
+        .tail(1)["index"]
+        .values
+    )
+
+    joined_df["last_epc_before_mcs"] = False
+    joined_df["last_epc_before_mcs"].iloc[last_epc_before_mcs_indices] = True
+
+    # Filter to either "last EPC before MCS" or "EPC after MCS",
+    # then group by installation and take first record -
+    # this will be the last EPC before MCS
+    # if it exists, otherwise the first EPC after MCS
+    filtered_data = (
+        joined_df.loc[joined_df["last_epc_before_mcs"] | ~joined_df["epc_before_mcs"]]
+        .groupby("original_mcs_index")
+        .head(1)
+        .reset_index(drop=True)
+        .drop(columns=["epc_before_mcs", "last_epc_before_mcs"])
+    )
+
+    return filtered_data
+
+
```

Comment:
Ended up not fully testing this script because I couldn't install recordlinkage. That's on my side, though. And I've tested this code in the past so I'm sure it runs fine. Will give it a full run again once we've integrated the final EPC data load.

Reply:
Sounds good - yep let's test once everything's on the same branch